### DISCUSSION TITLE
Documentation on Kevin

### DISCUSSION THREAD

**Discussion by BradKML** [SmartManoj/Kevin#134] [Q&A]
1. Could you edit the README for Docker + Kevin (assuming the worst through WSL)?
2. How can parameters be set (e.g. top_p, temperature) when using the Docker one-liner?
3. Could more details be provided regarding the use of open weight local providers like Ollama?
4. What about flexible provider routers like OpenRouter? Can it handle long responses?

**Comment by SmartManoj** [comment]
> 2\. How can parameters be set (e.g. top_p, temperature) when using the Docker one-liner?

By prepending with LLM_ and in caps
```
-e LLM_TOP_P=0.9
-e LLM_TEMPERATURE=0.0
```

**Reply by BradKML** [reply]
^ yeah this might not be clear since I need to trace where `LLM_TEMPERATURE` pops up in the codebase for example

**Reply by SmartManoj** [reply]
Here LLM is the group name. https://github.com/All-Hands-AI/OpenHands/blob/4a8bf3d2d0b3a032b16962b62f2cce65c4fe1850/openhands/core/config/utils.py#L30

**Reply by BradKML** [reply]
Yeah I can see it now, definitely asking for documentation tho (e.g. where it `TOP_P` and how it is all uppercase) https://github.com/All-Hands-AI/OpenHands/blob/4a8bf3d2d0b3a032b16962b62f2cce65c4fe1850/openhands/core/config/utils.py#L85

**Reply by SmartManoj** [reply]
converted [here](https://github.com/All-Hands-AI/OpenHands/blob/4a8bf3d2d0b3a032b16962b62f2cce65c4fe1850/openhands/core/config/utils.py#L50 )

**Reply by BradKML** [reply]
Obligated to say: when in doubt, document (even if it could become FAQ) cus it could come in handy sooner

**Comment by SmartManoj** [comment]
1) Updated to use the development guide. Docker takes long time to download, high storage space and high memory?
2)
3) Isn't [this Guide](https://docs.all-hands.dev/modules/usage/llms/local-llms) not enough?

**Reply by BradKML** [reply]
For 2. didn't notice the connection between the `-e` flag with `config.toml`, might need a little bit of elaboration as to what goes where

**Reply by SmartManoj** [reply]
![image](https://github.com/user-attachments/assets/c86cfe49-42c7-4af2-89a8-36a466f06b8d)
![image](https://github.com/user-attachments/assets/dc2b98bf-fd8a-486f-8447-97b6a2ce3344)
Number 3 changed to 2.

**Comment by SmartManoj** [comment]
4. Yes. Did you get any error?

**Reply by BradKML** [reply]
YEP (sadly) https://www.all-hands.dev/share?share_id=448cb24f9a0b8266d38e4a6ec46377c0163de390f30c6feedbeafa3253c3aeab

**Reply by SmartManoj** [reply]
Could you share the error also?

**Reply by SmartManoj** [reply]
Using `groq/gemma2-9b-it`, https://www.all-hands.dev/share?share_id=9beea1e970538dbe8c1285e0aeea21f959a5fb930cab47c21f3935818712c6fe

**Reply by BradKML** [reply]
I lost the last one (stuck in the loop error due to doing the same instruction and failing to match on OpenHands) but I am doing another run. And yes it seems to be very common to fail at string replacements. https://www.all-hands.dev/share?share_id=aab20962aaf38a65b94cca9507d1e615a611c7ca809c061de1f55d669401ef4b https://pastebin.com/LMfzHeE2

**Reply by SmartManoj** [reply]
Which model did you use?

Using `openrouter/google/gemma-2-9b-it:free` (34 steps) [Wrote 1038 words]
https://www.all-hands.dev/share?share_id=4f75761fd756e226cf6c135a552ff3455f6eba582b21d08d14c6b8495a643648

**Reply by SmartManoj** [reply]
Root cause. It didn't mention the number of remaining lines.

![image](https://github.com/user-attachments/assets/3ac5cd8f-b22d-4732-8831-f8a1e5eeac3a)

---
Tool name `str_replace_editor` is replaced with `file_editor`  [here](https://github.com/OpenDevin/OpenDevin/tree/main/openhands/agenthub/codeact_agent/function_calling.py#L507) -> may confuse LLM.

![image](https://github.com/user-attachments/assets/9fd3e8c0-07a1-477c-bf3a-346f3fd7da80)
`old_str` is mentioned as required [here](https://github.com/OpenDevin/OpenDevin/tree/main/openhands/agenthub/codeact_agent/function_calling.py#L249-L256).

---

Which one do you prefer?

line_number before content
```python
print(file_editor(**{'command': 'insert', 'path': '/workspace/backtesting_architectures.md', 'new_str': 'dummy', 'insert_line': 50}))
```
line_number after content
```python
insert_content_after_line(file_name='/workspace/a7.md', line_number=179, content='''
dummy
 ''')
```

**Reply by BradKML** [reply]
@SmartManoj CodeAct agent (default) with QwenCoder 2.5 (could use a more normal one like Qwen proper or DeepSeek). Using 0.16 Docker image. This is more geared towards testing if it is broken on OpenRouter side. Will check with Qwen proper with 0.17 to see if it replicates.

Edit: yeah same issue of stuck on loop with QwenCoder 2.5 on 0.17, probs switching models to see if anything else would happen https://www.all-hands.dev/share?share_id=ca614e1a43935e7169f6580cd0bb532d98578f92b74a321dd61d82cbe7a08633 https://pastebin.com/LssVZNin

**Reply by SmartManoj** [reply]
No problem from OpenRouter side. 

TLDR of the previous comment: The agent didn't mention the number of remaining lines.

**Reply by BradKML** [reply]
Do you think this is prompt side or agent-side? Because they should be keeping track of the word count, but at the same time the loop is stuck because of other issues (article containing repetitions, can't delete text to make things coherent)

**Reply by SmartManoj** [reply]
```python
print(file_editor(**{'command': 'insert', 'path': '/workspace/backtesting_architectures.md', ...
```

What do you think the agent should return for this Action?

**Reply by BradKML** [reply]
Expected agent actions:
1. Either it initially appends as much text as possible (easy), and then edit away duplicate information (could be bugged), OR
2. It creates a comprehensive outline for the text (hard planning), and then proceed to append information more carefully (easy)

**Reply by SmartManoj** [reply]
After inserted, the agent should simply say "Inserted successfully" but instead of that it shows half content and the following hint.
```
<response clipped><NOTE>To save on context only part of this file has been shown to you. You should retry this tool after you have searched inside the file with `grep -n` in order to find the line numbers of what you are looking for.</NOTE>.
```

Here the hint is wrong. This is applicable for the `view` command and not for the `insert` command.
Also, there should be a line break before the hint for humans.

![image](https://github.com/user-attachments/assets/59efd487-7644-4c07-9d68-e480afc3fa13)

**Reply by BradKML** [reply]
That is weird then, since the AI should be able to read the message as successful insertion and be done with it (if it fails it should just throw an error or display an exception). Checking is another step altogether.

**Reply by SmartManoj** [reply]
Yes.
